lib/Basics/async.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include "expected.h"
+#include <coroutine>
+#include <atomic>
+#include <stdexcept>
+#include <utility>
+#include <source_location>
+
+namespace arangodb {
+
+template<typename T>
+struct async_promise;
+template<typename T>
+struct async;
+
+template<typename T>
+struct async_promise_base {
+  using promise_type = async_promise<T>;
+
+  std::suspend_never initial_suspend() noexcept { return {}; }
+  auto final_suspend() noexcept {
+    struct awaitable {
+      bool await_ready() noexcept { return false; }
+      std::coroutine_handle<> await_suspend(
+          std::coroutine_handle<> self) noexcept {
+        auto addr =
+            _promise->_continuation.exchange(std::noop_coroutine().address());
+        if (addr == nullptr) {
+          return std::noop_coroutine();
+        } else {
+          auto continuation = std::coroutine_handle<>::from_address(addr);
+          if (continuation == std::noop_coroutine()) {
+            self.destroy();
+          }
```

Comment:
```suggestion
          if (continuation == std::noop_coroutine()) {
            self.destroy();
          }
```
From https://en.cppreference.com/w/cpp/coroutine/noop_coroutine:
> Return values from different calls to noop_coroutine may and may not compare equal. 
and
> If there has been already a coroutine state of no-op coroutine, it is unspecified whether a subsequent call to noop_coroutine returns a previously obtained coroutine handle, or a coroutine handle referring to a new coroutine state of no-op coroutine.

I think if you can't use `std::noop_coroutine().address()` as a special value, but need to assign e.g. a `static auto noop_coroutine = std::noop_coroutine();` to get at stable address.